Buildings/Controls/OBC/CDL/Logical/VariablePulse.mo

```diff
@@ -0,0 +1,329 @@
+within Buildings.Controls.OBC.CDL.Logical;
+block VariablePulse
+  "Generate boolean pulse with the width specified by input"
+
+  parameter Real samplePeriod(unit="s")
+    "Sample period of the width input";
+  parameter Real period(unit="s")
+    "Time for one pulse period";
+  parameter Real chaWidThr=0.05
+    "Threshold for checking if the width input is changed. It is the ratio of the value change to the original value"
+    annotation (Dialog(tab="Advanced"));
+  parameter Real zerWidThr=0.01
+    "Threshold for checking if the width input is greater than zero"
+    annotation (Dialog(tab="Advanced"));
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput uWid(
+    final min=0,
+    final max=1,
+    final unit="1")
+    "Ratio of the period that the output should be true"
+    annotation (Placement(transformation(extent={{-180,-20},{-140,20}}),
+        iconTransformation(extent={{-140,-20},{-100,20}})));
+  Buildings.Controls.OBC.CDL.Interfaces.BooleanOutput y
+    "Boolean pulse when the input is greater than zero"
+    annotation (Placement(transformation(extent={{140,-130},{180,-90}}),
+        iconTransformation(extent={{100,-20},{140,20}})));
+
+protected
+  Buildings.Controls.OBC.CDL.Discrete.Sampler sam(
```

Comment:
I don't think this should be time sampler. Shouldn't it rather be triggered whenever the downstream `greThr` triggers?

Reply:
Thanks for the suggestion. I replaced the sampler with the triggered one, with the `pre` to break to loop.